WeatherApp-UIKit/WeatherApp-UIKit/WeatherViewModelTests.swift

```diff
@@ -0,0 +1,196 @@
+    let location = locationObservable(with: [
+      .init(time: 100, value: .next(.location(mockLocations[0]))),
+      .init(time: 200, value: .next(.location(mockLocations[1]))),
+      .init(time: 300, value: .next(.location(mockLocations[2]))),
+    ])
+
+    var recordedLocations: [Coordinates] = []
+    let api: WeatherViewModel.API = {
+      recordedLocations.append($0)
+      return .just(.mock)
+    }
+
+    let sut = WeatherViewModel(weatherAPI: api, currentLocation: location)
+    let result = scheduler.createObserver(WeatherViewModelState.self)
+    sut.state.drive(result).disposed(by: disposeBag)
+
+    scheduler.start()
+
+    XCTAssertEqual(recordedLocations, mockLocations)
+  }
+
+  func testErrorIsPropagated() {
+    let location = locationObservable(with: [
+      .init(time: 0, value: .next(.loading)),
+      .init(time: 100, value: .next(.error(MockError())))
+    ])
+
+    let sut = WeatherViewModel(weatherAPI: mockApi, locale: locale, currentLocation: location)
+    let result = runAndObserve(sut.state)
+
+    XCTAssertEqual(result.events, [
+      .init(time: 0, value: .next(.loading)),
+      .init(time: 100, value: .next(.error)),
+    ])
+  }
+
+  func testLoadingIsPropagated() {
+    let location = locationObservable(with: [
+      .init(time: 0, value: .next(.location(mockLocations[0])))
+    ])
+
+    let api: WeatherViewModel.API = { _ in
+      self.scheduler
+        .createColdObservable([.next(500, Weather.mock), .completed(500)])
+        .asObservable()
+    }
+
+    let sut = WeatherViewModel(weatherAPI: api, locale: locale, currentLocation: location)
+    let result = runAndObserve(sut.state)
+
+    XCTAssertEqual(result.events,
+      [
+        .init(time: 0, value: .next(.loading)),
+        .init(time: 500, value: .next(.loadedMock))
+      ]
+    )
+  }
+
+  func testSuccess() {
+    let location = locationObservable(with: [
+      .init(time: 0, value: .next(.loading)),
+      .init(time: 100, value: .next(.location(mockLocations[0])))
+    ])
+
+    let api: WeatherViewModel.API = { _ in
+      self.scheduler
+        .createColdObservable([.init(time: 100, value: .next(Weather.mock))])
+        .asObservable()
+    }
+
+    let sut = WeatherViewModel(weatherAPI: api, locale: locale, currentLocation: location)
+    let result = runAndObserve(sut.state)
+
+    XCTAssertEqual(result.events, [
+      .init(time: 0, value: .next(.loading)),
+      .init(time: 100, value: .next(.loading)),
+      .init(time: 200, value: .next(.loadedMock))
+    ])
+  }
+
+  func testLikeDislikeButtonsPress() {
+    let location = locationObservable(with: [
+      .init(time: 0, value: .next(.location(mockLocations[0])))
+    ])
+
+    let api: WeatherViewModel.API = { _ in .just(.mock) }
+
+    var likedWeather: Weather?
+    var dislikedWeather: Weather?
+
+    let storage: WeatherViewModel.Storage = { _weather, liked in
+      if liked { likedWeather = _weather } else { dislikedWeather = _weather }
```

Comment:
can we write it on more lines?

Reply:
![](https://vojtastavik.com/images/2016-03-17/custom-array-5.gif)